yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/launcher/DockerComputerIOLauncher.java

```diff
@@ -205,6 +205,7 @@ public void onNext(Frame item) {
             } else if (item.getStreamType() == StreamType.STDOUT) {
                 try {
                     outputStream.write(item.getPayload());
+                    outputStream.flush();
```

Comment:
Wouldn't this close stream while it should be ready for next reads?

Reply:
No, the problem was that the buffer in piped streams did not get transmitted and the execution of commands took forever till some other timeout happend. There is still a problem with some closing of the stream but I am currently not sure where it comes from. This only happens if the node is idle. You can reproduce it, if you select the cloud retention and reuse the node. After some idle time, the pipe gets disconnected.

Without flush:

```
[Pipeline] timestamps
[Pipeline] {
[Pipeline] stage
[Pipeline] { (test)
[Pipeline] parallel
[Pipeline] [default-io] { (Branch: default-io)
[Pipeline] [default-io] stage
[Pipeline] [default-io] { (default-io)
[Pipeline] [default-io] node
00:00:18.759 [default-io] Still waiting to schedule task
00:00:18.765 [default-io] All nodes of label ‘default’ are offline
00:00:42.662 [default-io] Running on local-5eb5cf683fa0 in /home/jenkins/agent/workspace/test
[Pipeline] [default-io] {
[Pipeline] [default-io] sh
00:00:46.466 [default-io] [test] Running shell script
00:00:46.969 [default-io] + uname -a
00:00:46.970 [default-io] Linux 5eb5cf683fa0 4.13.0-16-lowlatency #19-Ubuntu SMP PREEMPT Wed Oct 11 19:51:52 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
[Pipeline] [default-io] sh
00:00:47.069 [default-io] [test] Running shell script
00:00:47.335 [default-io] + ls -al
00:00:47.336 [default-io] total 0
00:00:47.336 [default-io] drwxr-xr-x 1 root root  0 Nov 21 15:05 .
00:00:47.336 [default-io] drwxr-xr-x 1 root root 24 Nov 21 15:05 ..
[Pipeline] [default-io] }
[Pipeline] [default-io] // node
[Pipeline] [default-io] }
[Pipeline] [default-io] // stage
[Pipeline] [default-io] }
[Pipeline] // parallel
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // timestamps
[Pipeline] End of Pipeline
```

With flush:

```
[Pipeline] timestamps
[Pipeline] {
[Pipeline] stage
[Pipeline] { (test)
[Pipeline] parallel
[Pipeline] [default-io] { (Branch: default-io)
[Pipeline] [default-io] stage
[Pipeline] [default-io] { (default-io)
[Pipeline] [default-io] node
00:00:04.801 [default-io] Running on local-f24951ec9bc9 in /home/jenkins/agent/workspace/test
[Pipeline] [default-io] {
[Pipeline] [default-io] sh
00:00:04.891 [default-io] [test] Running shell script
00:00:05.158 [default-io] + uname -a
00:00:05.159 [default-io] Linux f24951ec9bc9 4.13.0-16-lowlatency #19-Ubuntu SMP PREEMPT Wed Oct 11 19:51:52 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
[Pipeline] [default-io] sh
00:00:05.227 [default-io] [test] Running shell script
00:00:05.506 [default-io] + ls -al
00:00:05.506 [default-io] total 0
00:00:05.506 [default-io] drwxr-xr-x 1 root root  0 Nov 21 15:10 .
00:00:05.506 [default-io] drwxr-xr-x 1 root root 24 Nov 21 15:10 ..
[Pipeline] [default-io] }
[Pipeline] [default-io] // node
[Pipeline] [default-io] }
[Pipeline] [default-io] // stage
[Pipeline] [default-io] }
[Pipeline] // parallel
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // timestamps
[Pipeline] End of Pipeline
```

Reply:
The steps do not return their results immediatly but wait for something else to fill the buffer. This delays the execution.

Reply:
What connector do you use?

Reply:
I am using netty, did not do much other configuration. Just added the ip of my docker host and created the templates.

java path and -Xmx1G are the only options I am passing to the slave to get my images running.

Reply:
btw, iolauncher was fixed only in some latest versions https://github.com/jenkinsci/jenkins/pull/3067/commits/1f476da2e1daf24edfa8431769fb59c73a0ad088

Reply:
just updated to 2.91 but still the same issue with idle disconnect ... will try to investigate further ...

Reply:
navigate to slave page and open "log" page for this slave. It should print startup attempts. Then slave will be deleted, but page will collect messages.

Reply:
```
Attaching to container...
<===[JENKINS REMOTING CAPACITY]===>ERROR: channel started

Remoting version: 3.14
This is a Unix agent
Evacuated stdout
Agent successfully connected and online
ERROR: Connection terminated
java.io.IOException: Read end dead
	at java.io.PipedInputStream.checkStateForReceive(PipedInputStream.java:262)
	at java.io.PipedInputStream.receive(PipedInputStream.java:226)
	at java.io.PipedOutputStream.write(PipedOutputStream.java:149)
	at java.io.OutputStream.write(OutputStream.java:75)
	at hudson.remoting.ChunkedOutputStream.sendFrame(ChunkedOutputStream.java:89)
	at hudson.remoting.ChunkedOutputStream.sendBreak(ChunkedOutputStream.java:62)
	at hudson.remoting.ChunkedCommandTransport.writeBlock(ChunkedCommandTransport.java:46)
	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.write(AbstractSynchronousByteArrayCommandTransport.java:45)
	at hudson.remoting.Channel.send(Channel.java:675)
	at hudson.remoting.Channel.close(Channel.java:1391)
	at hudson.remoting.Channel.close(Channel.java:1358)
	at hudson.slaves.SlaveComputer.closeChannel(SlaveComputer.java:737)
	at hudson.slaves.SlaveComputer.access$800(SlaveComputer.java:96)
	at hudson.slaves.SlaveComputer$3.run(SlaveComputer.java:655)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Reply:
```
A thread (Thread-8/254) died unexpectedly due to an uncaught exception, this may leave your Jenkins in a bad way and is usually indicative of a bug in the code.
java.io.IOException: Pipe broken
	at java.io.PipedInputStream.read(PipedInputStream.java:321)
	at java.io.PipedInputStream.read(PipedInputStream.java:377)
	at java.io.InputStream.read(InputStream.java:101)
	at com.github.kostyasha.yad_docker_java.com.github.dockerjava.netty.InvocationBuilder$6.read(InvocationBuilder.java:308)
Caused: java.lang.RuntimeException
	at com.github.kostyasha.yad_docker_java.com.github.dockerjava.netty.InvocationBuilder$6.read(InvocationBuilder.java:310)
	at com.github.kostyasha.yad_docker_java.com.github.dockerjava.netty.InvocationBuilder$6.run(InvocationBuilder.java:320)
	at java.lang.Thread.run(Thread.java:748)
```

Reply:
Btw, could you try it 1) with freestyle job, 2) with jnlp launcher (to ensure that workflow works fine)?

Reply:
Sorry for the late answer, yes, both still work. When using freestyle job with the docker-io launcher, it suffers the same problem with broken pipe after some time.